File: dotnet/src/Sys.cs
// Process-level helpers: environment, home/temp directories, output routing and
// exit handling that mirror Node's `process` / `os` behavior.

using System.Runtime.InteropServices;
using System.Text;

namespace Skills;

internal static partial class Sys
{
    private static readonly object OutLock = new();
    private static volatile bool _redirectStdout;
    private static int _exitCode;
    private static readonly List<Action> ExitHooks = new();

    // ─── Output ───

    /// When set, everything written through <see cref="Out"/> goes to stderr. This
    /// is how `add --json` keeps stdout reserved for the final JSON value.
    public static bool StdoutRedirected
    {
        get => _redirectStdout;
        set => _redirectStdout = value;
    }

    private static readonly Lazy<Stream> StdoutStream = new(Console.OpenStandardOutput);
    private static readonly Lazy<Stream> StderrStream = new(Console.OpenStandardError);

    private static void WriteTo(bool stderr, string s)
    {
        if (s.Length == 0) return;
        lock (OutLock)
        {
            if (OperatingSystem.IsWindows() && WinConsole.TryWrite(stderr, s)) return;
            var stream = stderr ? StderrStream.Value : StdoutStream.Value;
            var bytes = Encoding.UTF8.GetBytes(s);
            try
            {
                stream.Write(bytes, 0, bytes.Length);
                stream.Flush();
            }
            catch (IOException)
            {
                // Broken pipe: ignore, like Node does for EPIPE on exit.
            }
        }
    }

    /// Write to (possibly redirected) stdout.
    public static void Out(string s) => WriteTo(StdoutRedirected, s);

    public static void OutLine(string s = "") => Out(s + "\n");

    /// Write to the real stdout, bypassing redirection.
    public static void RealStdout(string s) => WriteTo(false, s);

    public static void Err(string s) => WriteTo(true, s);

    public static void ErrLine(string s = "") => Err(s + "\n");

    // ─── Exit ───

    public static int ExitCode
    {
        get => Volatile.Read(ref _exitCode);
        set => Volatile.Write(ref _exitCode, value);
    }

    /// Register a hook that runs before the process exits via <see cref="Exit"/>.
    public static void OnExit(Action hook)
    {
        lock (ExitHooks) ExitHooks.Add(hook);
    }

    public static void ClearExitHooks()
    {
        lock (ExitHooks) ExitHooks.Clear();
    }

    /// Equivalent of `process.exit(code)`: runs exit hooks, then terminates
    /// immediately (pending telemetry is intentionally not awaited, as in Node).
    [System.Diagnostics.CodeAnalysis.DoesNotReturn]
    public static void Exit(int code)
    {
        List<Action> hooks;
        lock (ExitHooks)
        {
            hooks = new List<Action>(ExitHooks);
            ExitHooks.Clear();
        }
        foreach (var h in hooks) h();
        Ui.OnProcessExit(code);
        Ui.RestoreTerminal();
        Environment.Exit(code);
        throw new InvalidOperationException("unreachable");
    }

    // ─── Environment ───

    /// `process.env.NAME` when set and non-empty (JS truthiness).
    public static string? Env(string name)
    {
        var v = Environment.GetEnvironmentVariable(name);
        return string.IsNullOrEmpty(v) ? null : v;
    }

    /// `process.env.NAME`, even if empty.
    public static string? EnvRaw(string name) => Environment.GetEnvironmentVariable(name);

    public static bool EnvTruthy(string name) => Env(name) != null;

    /// `process.env.NAME?.trim() || fallback`
    public static string? EnvTrimmed(string name)
    {
        var v = EnvRaw(name)?.Trim();
        return string.IsNullOrEmpty(v) ? null : v;
    }

    public static string Cwd() => Directory.GetCurrentDirectory();

    /// `os.homedir()` (libuv `uv_os_homedir`).
    public static string HomeDir()
    {
        if (OperatingSystem.IsWindows())
        {
            var p = Env("USERPROFILE");
            if (p != null) return p;
            var d = Env("HOMEDRIVE");
            var hp = Env("HOMEPATH");
            if (d != null && hp != null) return d + hp;
            return Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        }
        return Env("HOME") ?? Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
    }

    /// `os.tmpdir()`
    public static string TmpDir()
    {
        if (OperatingSystem.IsWindows())
        {
            var p = Env("TEMP") ?? Env("TMP") ?? ((Env("SystemRoot") ?? Env("windir") ?? "C:\\Windows") + "\\temp");
            if (p.Length > 1 && p.EndsWith('\\') && !p.EndsWith(":\\")) p = p[..^1];
            return p;
        }
        var t = Env("TMPDIR") ?? Env("TMP") ?? Env("TEMP") ?? "/tmp";
        if (t.Length > 1 && t.EndsWith('/')) t = t[..^1];
        return t;
    }

    private const string TempChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    /// `fs.mkdtemp(join(tmpdir(), prefix))`
    public static string MkdTemp(string prefix)
    {
        var baseDir = TmpDir();
        Directory.CreateDirectory(baseDir);
        for (var attempt = 0; attempt < 100; attempt++)
        {
            var suffix = new char[6];
            for (var i = 0; i < 6; i++) suffix[i] = TempChars[Random.Shared.Next(TempChars.Length)];
            var dir = NodePath.Join(baseDir, prefix + new string(suffix));
            if (Directory.Exists(dir) || File.Exists(dir)) continue;
            Directory.CreateDirectory(dir);
            return dir;
        }
        throw new IOException("failed to create temp dir");
    }

    public static bool StdinIsTty() => !Console.IsInputRedirected;

    public static bool StdoutIsTty() => !Console.IsOutputRedirected;

    public static int? TerminalColumns()
    {
        if (!StdoutIsTty()) return null;
        try
        {
            var w = Console.WindowWidth;
            return w > 0 ? w : null;
        }
        catch
        {
            return null;
        }
    }

    public static int? TerminalRows()
    {
        if (!StdoutIsTty()) return null;
        try
        {
            var h = Console.WindowHeight;
            return h > 0 ? h : null;
        }
        catch
        {
            return null;
        }
    }

    /// `new Date().toISOString()`
    public static string NowIso() => DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'", System.Globalization.CultureInfo.InvariantCulture);

    /// Enable ANSI escape processing on the Windows console (Node's libuv
    /// translates escapes itself, so output must render the same way).
    public static void EnableVirtualTerminal()
    {
        if (OperatingSystem.IsWindows()) WinConsole.EnableVt();
    }

    private static partial class WinConsole
    {
        private const int StdOutputHandle = -11;
        private const int StdErrorHandle = -12;
        private const uint EnableVirtualTerminalProcessing = 0x0004;

        [LibraryImport("kernel32.dll", SetLastError = true)]
        private static partial IntPtr GetStdHandle(int nStdHandle);

        [LibraryImport("kernel32.dll", SetLastError = true)]
        [return: MarshalAs(UnmanagedType.Bool)]
        private static partial bool GetConsoleMode(IntPtr hConsoleHandle, out uint lpMode);

        [LibraryImport("kernel32.dll", SetLastError = true)]
        [return: MarshalAs(UnmanagedType.Bool)]
        private static partial bool SetConsoleMode(IntPtr hConsoleHandle, uint dwMode);

        [LibraryImport("kernel32.dll", SetLastError = true, StringMarshalling = StringMarshalling.Utf16)]
        [return: MarshalAs(UnmanagedType.Bool)]
        private static partial bool WriteConsoleW(IntPtr hConsoleOutput, string lpBuffer, uint nNumberOfCharsToWrite, out uint lpNumberOfCharsWritten, IntPtr lpReserved);

        private static readonly Lazy<IntPtr?> OutHandle = new(() => ConsoleHandle(StdOutputHandle));
        private static readonly Lazy<IntPtr?> ErrHandle = new(() => ConsoleHandle(StdErrorHandle));

        private static IntPtr? ConsoleHandle(int std)
        {
            var h = GetStdHandle(std);
            if (h == IntPtr.Zero || h == new IntPtr(-1)) return null;
            return GetConsoleMode(h, out _) ? h : null;
        }

        /// Write UTF-16 text directly to a console handle; false when the stream
        /// is not a console (pipe or file).
        public static bool TryWrite(bool stderr, string s)
        {
            var h = stderr ? ErrHandle.Value : OutHandle.Value;
            if (h == null) return false;
            var offset = 0;
            while (offset < s.Length)
            {
                var chunk = s.Substring(offset, Math.Min(8192, s.Length - offset));
                if (!WriteConsoleW(h.Value, chunk, (uint)chunk.Length, out var written, IntPtr.Zero) || written == 0) return true;
                offset += (int)written;
            }
            return true;
        }

        public static void EnableVt()
        {
            foreach (var h in new[] { OutHandle.Value, ErrHandle.Value })
            {
                if (h is { } handle && GetConsoleMode(handle, out var mode))
                    SetConsoleMode(handle, mode | EnableVirtualTerminalProcessing);
            }
        }
    }
}

File: rust/src/sys.rs
//! Process-level helpers: environment, home/temp directories, output routing
//! and exit handling that mirror Node's `process` / `os` behavior.

use std::io::{IsTerminal, Write};
use std::sync::atomic::{AtomicBool, AtomicI32, Ordering};
use std::sync::Mutex;

/// When set, everything written through [`write_out`] goes to stderr. This is
/// how `add --json` keeps stdout reserved for the final JSON value.
static REDIRECT_STDOUT: AtomicBool = AtomicBool::new(false);
static EXIT_CODE: AtomicI32 = AtomicI32::new(0);

type ExitHook = Box<dyn FnOnce() + Send>;
static EXIT_HOOKS: Mutex<Vec<ExitHook>> = Mutex::new(Vec::new());

pub fn set_stdout_redirect(on: bool) {
    REDIRECT_STDOUT.store(on, Ordering::SeqCst);
}

pub fn stdout_redirected() -> bool {
    REDIRECT_STDOUT.load(Ordering::SeqCst)
}

/// Write to (possibly redirected) stdout.
pub fn write_out(s: &str) {
    if stdout_redirected() {
        let mut err = std::io::stderr().lock();
        let _ = err.write_all(s.as_bytes());
        let _ = err.flush();
    } else {
        let mut out = std::io::stdout().lock();
        let _ = out.write_all(s.as_bytes());
        let _ = out.flush();
    }
}

/// Write directly to the real stdout, bypassing redirection.
pub fn write_real_stdout(s: &str) {
    let mut out = std::io::stdout().lock();
    let _ = out.write_all(s.as_bytes());
    let _ = out.flush();
}

pub fn write_err(s: &str) {
    let mut err = std::io::stderr().lock();
    let _ = err.write_all(s.as_bytes());
    let _ = err.flush();
}

#[macro_export]
macro_rules! outln {
    () => { $crate::sys::write_out("\n") };
    ($($arg:tt)*) => { $crate::sys::write_out(&format!("{}\n", format_args!($($arg)*))) };
}

#[macro_export]
macro_rules! out {
    ($($arg:tt)*) => { $crate::sys::write_out(&format!($($arg)*)) };
}

#[macro_export]
macro_rules! errln {
    () => { $crate::sys::write_err("\n") };
    ($($arg:tt)*) => { $crate::sys::write_err(&format!("{}\n", format_args!($($arg)*))) };
}

pub fn set_exit_code(code: i32) {
    EXIT_CODE.store(code, Ordering::SeqCst);
}

pub fn exit_code() -> i32 {
    EXIT_CODE.load(Ordering::SeqCst)
}

/// Register a hook that runs before the process exits via [`exit`].
pub fn on_exit(hook: ExitHook) {
    EXIT_HOOKS.lock().unwrap().push(hook);
}

pub fn clear_exit_hooks() {
    EXIT_HOOKS.lock().unwrap().clear();
}

/// Equivalent of `process.exit(code)`: runs exit hooks, then terminates
/// immediately (pending telemetry is intentionally not awaited, as in Node).
pub fn exit(code: i32) -> ! {
    let hooks: Vec<ExitHook> = std::mem::take(&mut *EXIT_HOOKS.lock().unwrap());
    for hook in hooks {
        hook();
    }
    crate::ui::on_process_exit(code);
    crate::ui::restore_terminal();
    let _ = std::io::stdout().flush();
    let _ = std::io::stderr().flush();
    std::process::exit(code);
}

/// `process.env.NAME` when set and non-empty (JS truthiness).
pub fn env(name: &str) -> Option<String> {
    match std::env::var(name) {
        Ok(v) if !v.is_empty() => Some(v),
        _ => None,
    }
}

/// `process.env.NAME` when set, even if empty.
pub fn env_raw(name: &str) -> Option<String> {
    std::env::var(name).ok()
}

pub fn env_truthy(name: &str) -> bool {
    env(name).is_some()
}

/// `process.env.NAME?.trim() || fallback`
pub fn env_trimmed(name: &str) -> Option<String> {
    env_raw(name)
        .map(|v| v.trim().to_string())
        .filter(|v| !v.is_empty())
}

pub fn cwd() -> String {
    std::env::current_dir()
        .map(|p| p.to_string_lossy().to_string())
        .unwrap_or_else(|_| ".".to_string())
}

/// `os.homedir()` (libuv `uv_os_homedir`).
pub fn homedir() -> String {
    #[cfg(windows)]
    {
        if let Some(p) = env("USERPROFILE") {
            return p;
        }
        if let (Some(d), Some(p)) = (env("HOMEDRIVE"), env("HOMEPATH")) {
            return format!("{}{}", d, p);
        }
        "C:\\".to_string()
    }
    #[cfg(not(windows))]
    {
        if let Some(p) = env("HOME") {
            return p;
        }
        "/".to_string()
    }
}

/// `os.tmpdir()`
pub fn tmpdir() -> String {
    #[cfg(windows)]
    {
        let mut p = env("TEMP").or_else(|| env("TMP")).unwrap_or_else(|| {
            format!(
                "{}\\temp",
                env("SystemRoot")
                    .or_else(|| env("windir"))
                    .unwrap_or_else(|| "C:\\Windows".into())
            )
        });
        if p.len() > 1 && p.ends_with('\\') && !p.ends_with(":\\") {
            p.pop();
        }
        p
    }
    #[cfg(not(windows))]
    {
        let mut p = env("TMPDIR")
            .or_else(|| env("TMP"))
            .or_else(|| env("TEMP"))
            .unwrap_or_else(|| "/tmp".into());
        if p.len() > 1 && p.ends_with('/') {
            p.pop();
        }
        p
    }
}

/// `fs.mkdtemp(join(tmpdir(), prefix))`
pub fn mkdtemp(prefix: &str) -> std::io::Result<String> {
    let base = tmpdir();
    std::fs::create_dir_all(&base)?;
    for _ in 0..100 {
        let suffix = random_suffix(6);
        let dir = crate::paths::join(&[base.as_str(), &format!("{}{}", prefix, suffix)]);
        match std::fs::create_dir(&dir) {
            Ok(()) => return Ok(dir),
            Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => continue,
            Err(e) => return Err(e),
        }
    }
    Err(std::io::Error::other("failed to create temp dir"))
}

fn random_suffix(len: usize) -> String {
    use std::collections::hash_map::RandomState;
    use std::hash::{BuildHasher, Hasher};
    const CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    let mut out = String::new();
    let mut h = RandomState::new().build_hasher();
    h.write_u128(
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos())
            .unwrap_or(0),
    );
    h.write_u32(std::process::id());
    let mut v = h.finish();
    for _ in 0..len {
        out.push(CHARS[(v % CHARS.len() as u64) as usize] as char);
        v /= CHARS.len() as u64;
        if v == 0 {
            v = RandomState::new().build_hasher().finish();
        }
    }
    out
}

pub fn stdin_is_tty() -> bool {
    std::io::stdin().is_terminal()
}

pub fn stdout_is_tty() -> bool {
    std::io::stdout().is_terminal()
}

pub fn terminal_columns() -> Option<u16> {
    if !stdout_is_tty() {
        return None;
    }
    crossterm::terminal::size()
        .ok()
        .map(|(c, _)| c)
        .filter(|c| *c > 0)
}

pub fn terminal_rows() -> Option<u16> {
    if !stdout_is_tty() {
        return None;
    }
    crossterm::terminal::size()
        .ok()
        .map(|(_, r)| r)
        .filter(|r| *r > 0)
}

pub fn now_iso() -> String {
    // `new Date().toISOString()` → 2024-01-02T03:04:05.678Z
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default();
    let millis = now.as_millis() as i64;
    let secs = millis.div_euclid(1000);
    let ms = millis.rem_euclid(1000);
    let days = secs.div_euclid(86400);
    let rem = secs.rem_euclid(86400);
    let (h, m, s) = (rem / 3600, (rem % 3600) / 60, rem % 60);
    // Civil-from-days (Howard Hinnant)
    let z = days + 719468;
    let era = z.div_euclid(146097);
    let doe = z - era * 146097;
    let yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let mo = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if mo <= 2 { y + 1 } else { y };
    format!(
        "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}.{:03}Z",
        y, mo, d, h, m, s, ms
    )
}

pub fn is_windows() -> bool {
    cfg!(windows)
}
